worker/charmdownloader/manifold.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charmdownloader
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/worker/v2"
+	"github.com/juju/worker/v2/dependency"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/base"
+	client "github.com/juju/juju/api/charmdownloader"
+)
+
+// ManifoldConfig describes the resources used by the charmdownloader worker.
+type ManifoldConfig struct {
+	APICallerName string
+	Logger        Logger
+
+	NewCharmDownloaderAPI func(base.APICaller) CharmDownloaderAPI
```

Comment:
Usually everything in the config is validated.  A comment on why we're not validating NewCharmDownloaderAPI would be helpful.

Reply:
I changed the validation code to inject the default client if one is not provided.

Reply:
Actually I had to roll this back as the manifold config is not a pointer so I can't patch it while validating. I did add a comment however explaining that it is an optional field and that the default client will be used if no c-tor is specified.